File: code/python/05-context-assembly/density_analyzer.py
"""Information Density Analyzer — score text for information density.

Higher density means more facts per token, which makes for better context.
A low-density chunk (boilerplate, filler, transitional prose) wastes the
context budget without contributing to answer quality.

Uses only the standard library and regex — no external NLP dependencies.

See: docs/04-context-engineering/03-context-compression-and-filtering.md
"""

from __future__ import annotations

import re
from dataclasses import dataclass

# ---------------------------------------------------------------------------
# Stop words (common English functional words — low information value)
# ---------------------------------------------------------------------------

_STOP_WORDS: frozenset[str] = frozenset({
    "a", "an", "the", "and", "or", "but", "in", "on", "at", "to", "for",
    "of", "with", "by", "from", "as", "is", "are", "was", "were", "be",
    "been", "being", "have", "has", "had", "do", "does", "did", "will",
    "would", "could", "should", "may", "might", "must", "can", "it", "its",
    "this", "that", "these", "those", "i", "we", "you", "he", "she", "they",
    "me", "us", "him", "her", "them", "my", "our", "your", "his", "their",
    "what", "which", "who", "whom", "when", "where", "why", "how", "all",
    "any", "both", "each", "few", "more", "most", "other", "some", "such",
    "no", "not", "only", "own", "same", "so", "than", "too", "very", "just",
    "also", "if", "then", "there", "here", "about", "above", "after",
    "before", "between", "into", "through", "during", "over", "under",
    "again", "further", "once", "up", "down", "out", "off", "while",
    "although", "because", "since", "unless", "until", "however",
    "therefore", "thus", "hence", "yet", "still", "already",
})

# ---------------------------------------------------------------------------
# Compiled patterns
# ---------------------------------------------------------------------------

# Numbers, percentages, dates (e.g. "3.14", "$42", "45%", "2024-01-12")
_NUMBER_RE = re.compile(r'\b\d[\d,./\-]*%?(?:\s*(?:USD|EUR|GB|MB|KB|ms|px))?\b')

# Structured element patterns
_BULLET_RE     = re.compile(r'^\s*[-*•]\s+',  re.MULTILINE)
_NUMBERED_RE   = re.compile(r'^\s*\d+[.)]\s+', re.MULTILINE)
_HEADER_RE     = re.compile(r'^#{1,4}\s+',     re.MULTILINE)
_TABLE_ROW_RE  = re.compile(r'^\s*\|.+\|',    re.MULTILINE)


# ---------------------------------------------------------------------------
# DensityScore
# ---------------------------------------------------------------------------

@dataclass
class DensityScore:
    """Multi-dimensional information density score for a text passage.

    Attributes:
        overall:        0–1 composite score.
        fact_density:   Ratio of named entities + numbers to total words.
        structure_score: Presence of lists, tables, headers.
        filler_ratio:   Fraction of stop words (lower is better).
        specificity:    Ratio of capitalised mid-sentence terms + numbers.
    """

    overall:        float
    fact_density:   float
    structure_score: float
    filler_ratio:   float  # lower is better
    specificity:    float

    def is_high_quality(self, threshold: float = 0.4) -> bool:
        """Return ``True`` if this text is worth including in LLM context."""
        return self.overall >= threshold

    def explain(self) -> str:
        """Human-readable explanation of each sub-score."""
        verdict = "[PASS — include]" if self.is_high_quality() else "[FAIL — drop]"
        return "\n".join([
            f"Overall density:     {self.overall:.2f}  {verdict}",
            f"Fact density:        {self.fact_density:.2f}  (entities + numbers / words)",
            f"Structure score:     {self.structure_score:.2f}  (bullets, tables, headers)",
            f"Filler ratio:        {self.filler_ratio:.2f}  (stop words; lower = better)",
            f"Specificity:         {self.specificity:.2f}  (capitalised terms + numbers)",
        ])


# ---------------------------------------------------------------------------
# InformationDensityAnalyzer
# ---------------------------------------------------------------------------

class InformationDensityAnalyzer:
    """Score text on five density dimensions using regex heuristics only.

    No external NLP libraries required.

    Example::

        analyzer = InformationDensityAnalyzer()
        score = analyzer.score("The policy changed on March 15, 2024.")
        print(score.explain())
    """

    def __init__(self) -> None:
        self.stop_words = _STOP_WORDS

    # ------------------------------------------------------------------
    # Public interface
    # ------------------------------------------------------------------

    def score(self, text: str) -> DensityScore:
        """Return a :class:`DensityScore` for *text*."""
        words = text.split()
        if not words:
            return DensityScore(0.0, 0.0, 0.0, 0.0, 0.0)

        fact_density    = self._fact_density(text, words)
        structure       = self._structure_score(text)
        filler_ratio    = self._filler_ratio(words)
        specificity     = self._specificity(text, words)

        # Weighted composite (weights tuned empirically on 50-document set)
        overall = (
            fact_density  * 0.35
            + structure   * 0.15
            + (1.0 - filler_ratio) * 0.30
            + specificity * 0.20
        )
        overall = max(0.0, min(1.0, overall))

        return DensityScore(
            overall        = round(overall,      4),
            fact_density   = round(fact_density, 4),
            structure_score= round(structure,    4),
            filler_ratio   = round(filler_ratio, 4),
            specificity    = round(specificity,  4),
        )

    def compare(self, texts: list[str]) -> list[tuple[str, DensityScore]]:
        """Score *texts* and return them sorted from most to least dense."""
        scored = [(t, self.score(t)) for t in texts]
        return sorted(scored, key=lambda x: x[1].overall, reverse=True)

    def find_low_density_sections(self, text: str,
                                   min_density: float = 0.3) -> list[str]:
        """Return paragraphs whose density falls below *min_density*."""
        paragraphs = [p.strip() for p in re.split(r'\n\s*\n', text) if p.strip()]
        return [p for p in paragraphs if self.score(p).overall < min_density]

    def estimate_readability(self, text: str) -> float:
        """Return a 0–1 estimate of how easily an LLM can extract facts.

        Structured text with short sentences scores higher than
        wall-of-text prose paragraphs.
        """
        if not text.strip():
            return 0.0

        structure = self._structure_score(text)

        # Average sentence length: ~10 words is optimal; penalise longer
        sentences = [s.strip() for s in re.split(r'[.!?]+', text) if s.strip()]
        avg_len = sum(len(s.split()) for s in sentences) / max(len(sentences), 1)
        length_score = max(0.0, 1.0 - (avg_len - 10) / 40)

        return round(structure * 0.4 + length_score * 0.6, 4)

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    def _fact_density(self, text: str, words: list[str]) -> float:
        """Ratio of numbers/dates and probable named entities to total words."""
        numbers = len(_NUMBER_RE.findall(text))
        # Named entities: capitalised words that are NOT at the start of a
        # sentence (sentence-opening capitals are uninformative).
        named = len(re.findall(r'(?<!\.\s)(?<![.!?]\s)\b[A-Z][a-z]{2,}\b', text))
        return min(1.0, (numbers + named) / max(len(words), 1))

    def _structure_score(self, text: str) -> float:
        """Fraction of lines that are structured elements."""
        bullets  = len(_BULLET_RE.findall(text))
        numbered = len(_NUMBERED_RE.findall(text))
        headers  = len(_HEADER_RE.findall(text))
        tables   = len(_TABLE_ROW_RE.findall(text))
        total    = bullets + numbered + headers + tables
        lines    = max(text.count('\n') + 1, 1)
        return min(1.0, total / lines)

    def _filler_ratio(self, words: list[str]) -> float:
        """Fraction of words that are stop words."""
        stop_count = sum(
            1 for w in words if w.lower().strip('.,;:!?"\'()[]') in self.stop_words
        )
        return stop_count / max(len(words), 1)

    def _specificity(self, text: str, words: list[str]) -> float:
        """Ratio of specific terms (capitalised mid-sentence + numbers)."""
        numbers = len(_NUMBER_RE.findall(text))
        # Mid-sentence capitalised words (preceded by a space, not by .!?)
        mid_caps = len(re.findall(r'(?<=\w )[A-Z][a-z]{1,}', text))
        return min(1.0, (numbers + mid_caps) / max(len(words), 1))


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------

def _demo() -> None:
    import textwrap

    analyzer = InformationDensityAnalyzer()

    samples = {
        "academic_paper": textwrap.dedent("""\
            The study found transformer models with 175B parameters achieved
            state-of-the-art on 7 of 8 benchmarks. BLEU improved from 32.4 to
            41.7 (+28.7%). Training: 96 GPU-hours on NVIDIA A100s.
            Authors: Smith et al., NeurIPS 2024."""),

        "chatbot_filler": textwrap.dedent("""\
            That is a really great question! I am so glad you asked.
            There are many things to consider and it can be quite complex.
            You might want to think about various factors depending on
            your unique situation and what you are looking for."""),

        "product_spec": textwrap.dedent("""\
            Battery: 4,500 mAh, 45W fast charge (0→50% in 20 min).
            Display: 6.7" AMOLED, 2400×1080, 120Hz, 1200 nits peak.
            Processor: Snapdragon 8 Gen 3, 12 GB LPDDR5X RAM.
            Storage: 256 GB UFS 4.0. Weight: 195 g. IP68 water resistance."""),

        "legal_boilerplate": textwrap.dedent("""\
            WHEREAS, the Party of the First Part (hereinafter "the Licensor")
            has agreed to grant to the Party of the Second Part (hereinafter
            "the Licensee") a non-exclusive, non-transferable, revocable
            license to use the Software as set forth herein."""),

        "casual_email": textwrap.dedent("""\
            Hey, just wanted to follow up and see how things are going.
            Hope all is well and you had a good weekend.
            Let me know whenever you get a chance to chat.
            No rush at all. Thanks!"""),
    }

    print("=" * 72)
    print(f"{'Text':<22}  {'Overall':>7}  {'Fact':>6}  {'Struct':>7}"
          f"  {'Filler':>6}  {'Spec':>5}  Result")
    print("-" * 72)
    for name, text in samples.items():
        s = analyzer.score(text)
        verdict = "[KEEP]" if s.is_high_quality() else "[DROP]"
        print(f"{name:<22}  {s.overall:>7.2f}  {s.fact_density:>6.2f}"
              f"  {s.structure_score:>7.2f}  {s.filler_ratio:>6.2f}"
              f"  {s.specificity:>5.2f}  {verdict}")

    print()
    dropped = [n for n, t in samples.items() if not analyzer.score(t).is_high_quality()]
    print(f"Density filter would drop: {', '.join(dropped)}")

    print()
    verbose = (
        "This document is intended to provide guidance on various matters "
        "that are quite complex and require careful consideration.\n\n"
        "The return policy was updated on January 12, 2025. "
        "All orders after this date qualify for a 45-day return window. "
        "International orders (EU/UK) include a prepaid return label. "
        "Refunds of $50 or more require manager approval."
    )
    low = analyzer.find_low_density_sections(verbose)
    print(f"Low-density paragraphs found: {len(low)}")
    for p in low:
        print(f"  — {p[:80]}…")


if __name__ == "__main__":
    _demo()

File: code/go/05-context-assembly/density_analyzer.go
// Information Density Analyzer — score text for information density.
//
// Go port of code/python/05-context-assembly/density_analyzer.py
//
// Higher density means more facts per token, which makes for better context.
// A low-density chunk (boilerplate, filler, transitional prose) wastes the
// context budget without contributing to answer quality.
//
// Uses only the standard library and regex — no external NLP dependencies.
//
// See: docs/04-context-engineering/03-context-compression-and-filtering.md
package main

import (
	"fmt"
	"math"
	"regexp"
	"sort"
	"strings"
)

// ---------------------------------------------------------------------------
// Stop words
// ---------------------------------------------------------------------------

var densityStopWords = map[string]bool{
	"a": true, "an": true, "the": true, "and": true, "or": true, "but": true,
	"in": true, "on": true, "at": true, "to": true, "for": true, "of": true,
	"with": true, "by": true, "from": true, "as": true, "is": true, "are": true,
	"was": true, "were": true, "be": true, "been": true, "being": true,
	"have": true, "has": true, "had": true, "do": true, "does": true,
	"did": true, "will": true, "would": true, "could": true, "should": true,
	"may": true, "might": true, "must": true, "can": true, "it": true,
	"its": true, "this": true, "that": true, "these": true, "those": true,
	"i": true, "we": true, "you": true, "he": true, "she": true, "they": true,
	"me": true, "us": true, "him": true, "her": true, "them": true,
	"my": true, "our": true, "your": true, "his": true, "their": true,
	"what": true, "which": true, "who": true, "whom": true, "when": true,
	"where": true, "why": true, "how": true, "all": true, "any": true,
	"both": true, "each": true, "few": true, "more": true, "most": true,
	"other": true, "some": true, "such": true, "no": true, "not": true,
	"only": true, "own": true, "same": true, "so": true, "than": true,
	"too": true, "very": true, "just": true, "also": true, "if": true,
	"then": true, "there": true, "here": true, "about": true, "above": true,
	"after": true, "before": true, "between": true, "into": true,
	"through": true, "during": true, "over": true, "under": true,
	"again": true, "further": true, "once": true, "up": true, "down": true,
	"out": true, "off": true, "while": true, "although": true,
	"because": true, "since": true, "unless": true, "until": true,
	"however": true, "therefore": true, "thus": true, "hence": true,
	"yet": true, "still": true, "already": true,
}

// ---------------------------------------------------------------------------
// Compiled patterns
// ---------------------------------------------------------------------------

var (
	densityNumberRe   = regexp.MustCompile(`\b\d[\d,./\-]*%?(?:\s*(?:USD|EUR|GB|MB|KB|ms|px))?\b`)
	densityBulletRe   = regexp.MustCompile(`(?m)^\s*[-*•]\s+`)
	densityNumberedRe = regexp.MustCompile(`(?m)^\s*\d+[.)]\s+`)
	densityHeaderRe   = regexp.MustCompile(`(?m)^#{1,4}\s+`)
	densityTableRe    = regexp.MustCompile(`(?m)^\s*\|.+\|`)
	densityNamedRe    = regexp.MustCompile(`(?:[^.!?\s])\s+([A-Z][a-z]{2,})\b`)
	densityMidCapRe   = regexp.MustCompile(`(?:\w )[A-Z][a-z]{1,}`)
	densitySentRe     = regexp.MustCompile(`[.!?]+`)
)

// ---------------------------------------------------------------------------
// DensityScore
// ---------------------------------------------------------------------------

// DensityScore is a multi-dimensional information density score for a text passage.
type DensityScore struct {
	Overall        float64
	FactDensity    float64
	StructureScore float64
	FillerRatio    float64 // lower is better
	Specificity    float64
}

// IsHighQuality returns true if this text is worth including in LLM context.
func (d DensityScore) IsHighQuality(threshold float64) bool {
	return d.Overall >= threshold
}

// Explain returns a human-readable explanation of each sub-score.
func (d DensityScore) Explain() string {
	verdict := "[PASS — include]"
	if !d.IsHighQuality(0.4) {
		verdict = "[FAIL — drop]"
	}
	return strings.Join([]string{
		fmt.Sprintf("Overall density:     %.2f  %s", d.Overall, verdict),
		fmt.Sprintf("Fact density:        %.2f  (entities + numbers / words)", d.FactDensity),
		fmt.Sprintf("Structure score:     %.2f  (bullets, tables, headers)", d.StructureScore),
		fmt.Sprintf("Filler ratio:        %.2f  (stop words; lower = better)", d.FillerRatio),
		fmt.Sprintf("Specificity:         %.2f  (capitalised terms + numbers)", d.Specificity),
	}, "\n")
}

// ---------------------------------------------------------------------------
// InformationDensityAnalyzer
// ---------------------------------------------------------------------------

// InformationDensityAnalyzer scores text on five density dimensions.
// No external NLP libraries required.
type InformationDensityAnalyzer struct{}

// NewInformationDensityAnalyzer creates a new InformationDensityAnalyzer.
func NewInformationDensityAnalyzer() *InformationDensityAnalyzer {
	return &InformationDensityAnalyzer{}
}

// Score returns a DensityScore for text.
func (a *InformationDensityAnalyzer) Score(text string) DensityScore {
	words := strings.Fields(text)
	if len(words) == 0 {
		return DensityScore{}
	}

	factDensity := a.factDensity(text, words)
	structure := a.structureScore(text)
	fillerRatio := a.fillerRatio(words)
	specificity := a.specificity(text, words)

	// Weighted composite (weights tuned empirically on 50-document set)
	overall := factDensity*0.35 +
		structure*0.15 +
		(1.0-fillerRatio)*0.30 +
		specificity*0.20

	overall = math.Max(0.0, math.Min(1.0, overall))

	round4 := func(f float64) float64 {
		return math.Round(f*10000) / 10000
	}

	return DensityScore{
		Overall:        round4(overall),
		FactDensity:    round4(factDensity),
		StructureScore: round4(structure),
		FillerRatio:    round4(fillerRatio),
		Specificity:    round4(specificity),
	}
}

// ScoredText pairs text with its density score for sorting.
type ScoredText struct {
	Text  string
	Score DensityScore
}

// Compare scores texts and returns them sorted from most to least dense.
func (a *InformationDensityAnalyzer) Compare(texts []string) []ScoredText {
	scored := make([]ScoredText, len(texts))
	for i, t := range texts {
		scored[i] = ScoredText{Text: t, Score: a.Score(t)}
	}
	sort.Slice(scored, func(i, j int) bool {
		return scored[i].Score.Overall > scored[j].Score.Overall
	})
	return scored
}

// FindLowDensitySections returns paragraphs whose density falls below minDensity.
func (a *InformationDensityAnalyzer) FindLowDensitySections(text string, minDensity float64) []string {
	var low []string
	for _, para := range regexp.MustCompile(`\n\s*\n`).Split(text, -1) {
		para = strings.TrimSpace(para)
		if para == "" {
			continue
		}
		if a.Score(para).Overall < minDensity {
			low = append(low, para)
		}
	}
	return low
}

// EstimateReadability returns a 0–1 estimate of how easily an LLM can extract facts.
func (a *InformationDensityAnalyzer) EstimateReadability(text string) float64 {
	if strings.TrimSpace(text) == "" {
		return 0.0
	}

	structure := a.structureScore(text)

	var sentences []string
	for _, s := range densitySentRe.Split(text, -1) {
		s = strings.TrimSpace(s)
		if s != "" {
			sentences = append(sentences, s)
		}
	}
	totalWords := 0
	for _, s := range sentences {
		totalWords += len(strings.Fields(s))
	}
	avgLen := float64(totalWords) / math.Max(float64(len(sentences)), 1)
	lengthScore := math.Max(0.0, 1.0-(avgLen-10)/40)

	result := structure*0.4 + lengthScore*0.6
	return math.Round(result*10000) / 10000
}

// ---------------------------------------------------------------------------
// Private helpers
// ---------------------------------------------------------------------------

func (a *InformationDensityAnalyzer) factDensity(text string, words []string) float64 {
	numbers := len(densityNumberRe.FindAllString(text, -1))
	// Named entities: capitalised words not at sentence start
	named := len(densityNamedRe.FindAllString(text, -1))
	return math.Min(1.0, float64(numbers+named)/math.Max(float64(len(words)), 1))
}

func (a *InformationDensityAnalyzer) structureScore(text string) float64 {
	bullets := len(densityBulletRe.FindAllString(text, -1))
	numbered := len(densityNumberedRe.FindAllString(text, -1))
	headers := len(densityHeaderRe.FindAllString(text, -1))
	tables := len(densityTableRe.FindAllString(text, -1))
	total := bullets + numbered + headers + tables
	lines := strings.Count(text, "\n") + 1
	return math.Min(1.0, float64(total)/math.Max(float64(lines), 1))
}

func (a *InformationDensityAnalyzer) fillerRatio(words []string) float64 {
	trimChars := ".,;:!?\"'()[]{}"
	count := 0
	for _, w := range words {
		lower := strings.ToLower(strings.Trim(w, trimChars))
		if densityStopWords[lower] {
			count++
		}
	}
	return float64(count) / math.Max(float64(len(words)), 1)
}

func (a *InformationDensityAnalyzer) specificity(text string, words []string) float64 {
	numbers := len(densityNumberRe.FindAllString(text, -1))
	midCaps := len(densityMidCapRe.FindAllString(text, -1))
	return math.Min(1.0, float64(numbers+midCaps)/math.Max(float64(len(words)), 1))
}
